### PR TITLE
[Dynamic Protocol State] Protocol State Mutator

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/6802

### Context

This PR implements protocol state mutator. It's a dedicated component that will be used by compliance layer to create protocol state updater and later commit protocol state into storage. To maintain a correct index of `blockID` -> `protocol state` previously described logic has to be executed for EACH block.

### CODE REVIEW

File: state/protocol/protocol_state/mutator.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// CommitProtocolState commits the protocol state updater as part of DB transaction.
+// Has to be called for each block to correctly index the protocol state.
+// No errors are expected during normal operations.
+func (m *Mutator) CommitProtocolState(updater protocol.StateUpdater) func(tx *transaction.Tx) error {
+	return func(tx *transaction.Tx) error {
+		updatedState, updatedStateID, hasChanges := updater.Build()
+		if hasChanges {
+			err := m.protocolStateDB.StoreTx(updatedStateID, updatedState)(tx)
```
Comment (jordanschalm) [inline state/protocol/protocol_state/mutator.go:45]: Suppose we have two protocol state changes, $\delta_1$, $\delta_2$, and applying them sequentially to an initial state $\Delta$ results in $\Delta$ again.

$\Delta + \delta_1 = \Delta'$
$\Delta' + \delta_2 = \Delta$

For example, if the state changes subtract, then add, a constant value to `Weight`.

In that case, we will have an unhandled `storage.ErrAlreadyExists` error here, treated as irrecoverable. I don't think we should make the assumption at this layer that protocol states can never repeat, even if it happens to be true. 

**Suggestion:** Handle `ErrAlreadyExists` here, and consider as a no-op.

File: state/protocol/protocol_state/mutator_test.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	s.protocolStateDB = storagemock.NewProtocolState(s.T())
+	s.mutator = NewMutator(s.protocolStateDB)
+}
+
+// TestCreateUpdaterForUnknownBlock tests that CreateUpdater returns an error if the parent protocol state is not found.
+func (s *MutatorSuite) TestCreateUpdaterForUnknownBlock() {
+	candidate := unittest.BlockHeaderFixture()
+	s.protocolStateDB.On("ByBlockID", candidate.ParentID).Return(nil, storerr.ErrNotFound)
+	updater, err := s.mutator.CreateUpdater(candidate)
+	require.ErrorIs(s.T(), err, storerr.ErrNotFound)
```
Comment (jordanschalm) [inline state/protocol/protocol_state/mutator_test.go:36]: We should document this as an expected error, or wrap as `irrecoverable`.

**Comment by AlexHentschel** [comment]
Not sure if I mentioned this in one of the preceding PR reviews:
* I feel the goDoc of [interface `ProtocolState`](https://github.com/onflow/flow-go/blob/ddd1a1c485ad32e7547b604431d7d9ed696c1026/storage/protocol_state.go) could be extended
* Regarding https://github.com/onflow/flow-go/blob/ddd1a1c485ad32e7547b604431d7d9ed696c1026/storage/protocol_state.go#L10-L11
   If I understand correctly, the implementation does _not_ (contrary to the interface documentation) go through the caching layer, right? 


What do you think about extending the goDoc as follows:
```golang
// ProtocolState represents persistent storage for protocol state entries.
type ProtocolState interface {
	// StoreTx returns an anonymous function (intended to be executed as part of a badger transaction),
	// which persists the given protocol state as part of a DB tx.
	// In a nutshell, we want to maintain a map from `blockID` to `protocolStateID`. Upon call, the
	// anonymous function persists the specific map entry in the node's database.
	// Expected errors of the returned anonymous function:
	//   - storage.ErrAlreadyExists if a protocol state snapshot has already been stored under the given id.
	StoreTx(id flow.Identifier, protocolState *flow.ProtocolStateEntry) func(*transaction.Tx) error

	// Index returns an anonymous function that is intended to be executed as part of a database transaction.
	// In a nutshell, we want to maintain a map from `blockID` to `protocolStateID`. Upon call, the
	// anonymous function persists the specific map entry in the node's database.
	// Protocol convention:
	//   - Consider block B, whose ingestion might potentially lead to an updated protocol state. For example,
	//     the protocol state changes if we seal some execution results emitting service events.
	//   - For the key `blockID`, we use the identity of block B. As value, the hash of the resulting protocol
	//     state at the end of processing B is to be used.
	// Expected errors during normal operations:
	//   - storage.ErrAlreadyExists if the key already exists in the database.
	Index(blockID flow.Identifier, protocolStateID flow.Identifier) func(*transaction.Tx) error

	// ByID returns the protocol state by its ID.
	// Expected errors during normal operations:
	//  * `storage.ErrNotFound` if no protocol state snapshot with the given Identifier is known.
	ByID(id flow.Identifier) (*flow.RichProtocolStateEntry, error)

	// ByBlockID returns the protocol state by block ID.
	// Expected errors during normal operations:
	//  * `storage.ErrNotFound` if no snapshot of the protocol state has been indexed for the given block.
	ByBlockID(blockID flow.Identifier) (*flow.RichProtocolStateEntry, error)
}
```

File: state/protocol/protocol_state/mutator.go (hunk 1 above)
Comment (AlexHentschel) [reply state/protocol/protocol_state/mutator.go:45]: +1

We should certainly document possible error returns in detail for the [interface `ProtocolState`](https://github.com/onflow/flow-go/blob/ddd1a1c485ad32e7547b604431d7d9ed696c1026/storage/protocol_state.go). See [my comment below](https://github.com/onflow/flow-go/pull/4597#issuecomment-1709526474) for a detailed suggestion

File: state/protocol/protocol_state/mutator_test.go (hunk 2 above)
Comment (AlexHentschel) [reply state/protocol/protocol_state/mutator_test.go:36]: I think from the perspective of the `Mutator` the error should be benign, as it allows (under certain conditions, e.g. access node trying to serve an API request) to continue. 

See [my comment below](https://github.com/onflow/flow-go/pull/4597#issuecomment-1709526474) for a detailed suggestion on documenting error returns of [interface `ProtocolState`](https://github.com/onflow/flow-go/blob/ddd1a1c485ad32e7547b604431d7d9ed696c1026/storage/protocol_state.go).

**Review by AlexHentschel — APPROVED** [review]

File: state/protocol/protocol_state/mutator.go (hunk 1 above)
Comment (jordanschalm) [reply state/protocol/protocol_state/mutator.go:45]: ⚠️  Approving the PR - but please address this before merging.

File: state/protocol/protocol_state/mutator_test.go (hunk 2 above)
Comment (jordanschalm) [reply state/protocol/protocol_state/mutator_test.go:36]: ⚠️  Approving the PR - but please address this before merging.

**Review by jordanschalm — APPROVED** [review]
Commented on my outstanding points which we should address before merging.

File: state/protocol/protocol_state/mutator.go (hunk 1 above)
Comment (durkmurder) [reply state/protocol/protocol_state/mutator.go:47]: but in such case `hasChanges` has to be false, so it shouldn't happen. Nevertheless I will make it accept such case. [`7266c8f` (#4597)](https://github.com/onflow/flow-go/pull/4597/commits/7266c8f0adf0de4f3ac8fdc04ff2d7180f05b585)

File: state/protocol/protocol_state/mutator_test.go (hunk 2 above)
Comment (durkmurder) [reply state/protocol/protocol_state/mutator_test.go:36]: I've updated godoc so we expect `ErrNotFound` [`a1969a0` (#4597)](https://github.com/onflow/flow-go/pull/4597/commits/a1969a07843fd155d33e7f89a3c9e65136293261)

**Pull request merged**